ambersim/trajopt/cost.py

```diff
@@ -0,0 +1,173 @@
+from typing import Tuple
+
+import jax
+import jax.numpy as jnp
+from flax import struct
+
+from ambersim.trajopt.base import CostFunction, CostFunctionParams
+
+"""A collection of common cost functions."""
+
+
+class StaticGoalQuadraticCost(CostFunction):
+    """A quadratic cost function that penalizes the distance to a static goal.
+
+    This is the most vanilla possible quadratic cost. The cost matrices are static (defined at init time) and so is the
+    single, fixed goal. The gradient is as compressed as it can be in general (one matrix multiplication), but the
+    Hessian can be far more compressed by simplying referencing Q, Qf, and R - this implementation is inefficient and
+    dense.
+    """
+
+    def __init__(self, Q: jax.Array, Qf: jax.Array, R: jax.Array, qg: jax.Array, vg: jax.Array) -> None:
+        """Initializes a quadratic cost function.
+
+        Args:
+            Q (shape=(nx, nx)): The state cost matrix.
+            Qf (shape=(nx, nx)): The final state cost matrix.
+            R (shape=(nu, nu)): The control cost matrix.
+            qg (shape=(nq,)): The goal generalized coordinates.
+            vg (shape=(nv,)): The goal generalized velocities.
+        """
+        self.Q = Q
+        self.Qf = Qf
+        self.R = R
+        self.qg = qg
+        self.vg = vg
+
+    @staticmethod
+    def _setup_util(
+        qs: jax.Array, vs: jax.Array, qg: jax.Array, vg: jax.Array
+    ) -> Tuple[jax.Array, jax.Array, jax.Array, jax.Array, int, int, int]:
+        """Utility function that sets up the cost function.
+
+        Args:
+            qs (shape=(N + 1, nq)): The generalized positions over the trajectory.
+            vs (shape=(N + 1, nv)): The generalized velocities over the trajectory.
+            qg (shape=(nx,)): The goal generalized position.
+            vg (shape=(nv,)): The goal generalized velocity.
+
+        Returns:
+            xs (shape=(N + 1, nx)): The states over the trajectory.
+            xg (shape=(nx,)): The goal state.
+            xs_err (shape=(N, nx)): The state errors up to the final state.
+            xf_err (shape=(nx,)): The state error at the final state.
+            nq: The number of generalized coordinates.
+            nv: The number of generalized velocities.
+        """
+        xs = jnp.concatenate((qs, vs), axis=-1)
+        xg = jnp.concatenate((qg, vg), axis=-1)
+        xs_err = xs[:-1, :] - xg
+        xf_err = xs[-1, :] - xg
+        return xs, xg, xs_err, xf_err, qs.shape[-1], vs.shape[-1]
+
+    @staticmethod
+    def batch_quadform(bs: jax.Array, A: jax.Array) -> jax.Array:
+        """Computes a batched quadratic form for a single instance of A.
+
+        Args:
+            bs (shape=(..., n)): The batch of vectors.
+            A (shape=(n, n)): The matrix.
+
+        Returns:
+            val (shape=(...,)): The batch of quadratic forms.
+        """
+        return jnp.einsum("...i,ij,...j->...", bs, A, bs)
```

Comment:
btw, this is cool! I had no idea this existed

Reply:
I actually use it moderately frequently, but for this specific instance, you can thank Github copilot!